### PR TITLE
[SPARK-42441][CONNECT] Scala Client add Column APIs

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR adds most Column APIs for the Spark Connect Scala Client.

### Why are the changes needed?
We want the Scala Client to have API parity with the existing SparkSession/Dataset APIs.

### How was this patch tested?
Golden files, and I added a test for local behavior.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

File: connector/connect/common/src/test/resources/query-tests/queries/column_when_otherwise.proto.bin [hunk 1]
```diff
@@ -0,0 +1,14 @@
+�
+�Z��struct<id:bigint,a:int,b:double,d:struct<id:bigint,a:int,b:double>,e:array<int>,f:map<string,struct<id:bigint,a:int,b:double>>,g:string>OM
```
Comment (LuciferYang) [inline connector/connect/common/src/test/resources/query-tests/queries/column_when_otherwise.proto.bin:2]: What are these unrecognized characters?

Comment (hvanhovell) [reply connector/connect/common/src/test/resources/query-tests/queries/column_when_otherwise.proto.bin:2]: it is a binary file. It is not supposed to be readable.

**Comment by hvanhovell** [comment]
merging.

**Comment by HyukjinKwon** [comment]
FYI, bunch of Scala 2.13 tests fail (https://github.com/apache/spark/actions/runs/4187249065/jobs/7256849374#step:9:19944) should be fixed before cutting RC cc @xinrong-meng

**Comment by dongjoon-hyun** [comment]
Hi, @hvanhovell .
Since we are close to RC1 cut, could you test Scala 2.13 before merging, please?

**Pull request closed without merging**